### DISCUSSION TITLE
Add a section for storing tips data

### DISCUSSION THREAD

**Discussion by sst311212** [deividAlfa/stm32_soldering_iron_controller#83] [Ideas]
David, I think adding every tips after system reset or full erase is much annoying.
Then I add a new section for storing tip data, when profile resetting it will restore tips automatically.
We can write a small program to help modifying tips data, and it can also export tips data to computer for backup.

https://github.com/sst311212/stm32_soldering_iron_controller/tree/backup_tips

**Comment by deividAlfa** [comment]
I don't understand this?
A system reset doesn't erase the tips, only if you choose Reset All, Reset Profiles or Factory Reset with button.

Neither does updating the firmware, you should never erase the flash with the programmer.
If you full erase the flash, well, the backup will also be erased!

The profile will be erased automatically if the profile structure version changes, or if the checksum is bad.

There was bug that caused tips to be erased if the system settings checksum was bad, this was fixed.

Also, you removed 3KB flash from the program flash (124KB->121KB), using 1KB for tip backup?

- You're wasting space (2KB unused)
- STM32F072 has 2KB sectors, so it won't work there.
- You aren't saving the tips anywhere?
- Why add a restore function inside resetCurrentProfile, now it's impossible to reset the profile.
- How will you reset the profile then?
- I don't think 63 tips fit in just 1KB?
- Needs some CRC protection
- Needs sanity checks (Erased flash 0xFF will show 255 tips)

About the backup method, I though it several times, the easiest way would be to add a SD card slot, but as I always say, this firmware targets original boards, modifications will be used by a minority.

**Reply by sst311212** [reply]
> Neither does updating the firmware, you should never erase the flash with the programmer.

When I dirty upgrade firmware from v1.11.5b to v1.11.8 it will cause hard fault without full erase flash.

> Also, you removed 3KB flash from the program flash (124KB->121KB), using 1KB for tip backup?

128K - 1K(BACKUP) - 2K(TEMP) - 4K(GLOBAL) = 121K ?

> You aren't saving the tips anywhere?

Sure, but the pain is restore tips from Edit Tips menu manually.

> Why add a restore function inside resetCurrentProfile, now it's impossible to reset the profile.

I tested it already, it still can reset single and all profile.

> Needs some CRC protection

I defined the section just read only, and STM32 programmer will do checksum after downloading firmware isn't it?

**Reply by deividAlfa** [reply]
Ah! You actually found a bug! I had forgot to update the linker scripts when adding the new settings area!

**Comment by deividAlfa** [comment]
About the SD card, it could be easily done using FATFS and [kiwih's SPI SD driver](https://github.com/kiwih/cubeide-sd-card/tree/master/cubeide-sd-card/FATFS/Target), but KSGER V2 (The most common layout) has no SPI pins free, used for other stuff.
Still, it could be done with bit-banging... slow, but the data is small enough to not worry about.
However, the additional RAM usage by this is a concern in 10KB devices, it's very tight.

**Comment by deividAlfa** [comment]
Anyways, I've spent a crazy amount of hours fixing and enhancing everything (If you check the recent commits you'll get dizzy!)
I think the Hardfaults should be solved now.
Plus there's a lot more RAM free now, we actually could think on the SD card mod + .json / .ini file or something.

**Reply by sst311212** [reply]
I think my method is good, user can save his own tips into firmware directly, and it doesn't need any hardware mod, very convience.

It can also let user customize their desired pre-settings without compiling source code.

**Reply by deividAlfa** [reply]
I think we don't need this.
Tip structure has not changed for ages, so in the near future I will isolate them in their own section.
So when system or profile structure changes, the tips will remain.
And maybe a write-protect flag. (Edit: This can't be done in HW, Option Bits protect in 4KB blocks)
Something like this:
```
#                   [PROGRAM]  118KB

#                   [TIP BLOCK 1]     2KB ( Each block can hold 85 tips)
    TipCrc
    TipData
         structureVersion
         NumberOfTips
         Tips[TIP_NUM]

#                   [TIP BLOCK 2]     2KB
    TipCrc
    TipData
         structureVersion
         NumberOfTips
         Tips[TIP_NUM]

#                   [TIP BLOCK 3]     2KB
    TipCrc
    TipData
         structureVersion
         NumberOfTips
         Tips[TIP_NUM]

#                   [SETTINGS]  2KB
    SYSTEM
        SystemCRC
        SystemSettings
           structureVersion
           (settings...)

    PROFILES
        ProfileCRC[3]
        ProfileSettings[3]
           structureVersion
           (settings...)

    ADDONS
        AdonsCRC
        AdonsSettings
           structureVersion
           (settings...)

#                   [TEMP_SETTINGS] 2KB
    lastProfile
    lastTemp
    lastTip[3]
```

**Reply by deividAlfa** [reply]
@sst311212 So what do you think about the current tip structure?
It's now implemented.
They should no longer be erased by any means unless their own data is bad or wiped by user full reset.

**Reply by sst311212** [reply]
Tip structure is good, but everytime savetip will erase flash because checksum?  Can't it just use current tips number's area to do crc checksum or padding a dummy tip data to identify checksum border?

And temp setting will be erased when one of profile, tip[] and temp is end of use, I think we can use MSB to identify last data position.

**Reply by deividAlfa** [reply]
Yes it will, but I don't think it's problem, unless you modify tips 10K times, which is pretty much impossible.
Think it was a lot worse all this time, the entire settings had to be erased for any change.

Adding per-tip CRC? Perhabs... But I'm tired, people always forget I do this in my free time, and it was a lot of work, so for now I'm not changing anything else.

Temp setting has to be erased, it's the minimum page erase size.
Once a flash word is written, any further attempts to write data will result in error, only 0 can be written, so you can't write single bits until all are 0.
Still, it'll be erased every 128 tip/profile changes or 512 temperature changes, I think you're being a little paranoid about erasing the flash!

**Reply by sst311212** [reply]
I thought it can rewrite data to used flash, now I know why it should erase flash before storing new data. 😂

**Reply by deividAlfa** [reply]
Yeah you can write anything else than 0 only when the flash word is 0xFFFF (Erased).
That's why I had to use entire 16-bit values to store very small data like the profile or the tip, flash data is 16-bit, you can't write it in smaller chunks.

**Reply by sst311212** [reply]
16 bit for small data? but what about ansi string, it only use 8 bit per character.
So we can read per byte in flash, but needing write them in word size?

**Reply by deividAlfa** [reply]
Each flash word is 16 bit, you can't write smaller at a time.
That's why last tip/temperature/profile arrays are uint16_t, each new written value can only be 16 bit.

You might read STM32 RM008 manual.
